File: 33. Search in Rotated Sorted Array/33.go
// 33. Search in Rotated Sorted Array solution in GoLang
// https://leetcode.com/problems/search-in-rotated-sorted-array/description/

package leetcode

func search(nums []int, target int) int {
    start, end := 0, len(nums) - 1

    for start <= end {
        mid := start + ((end - start) / 2)

        if nums[mid] == target {
            return mid
        } else if nums[mid] >= nums[start] {
            if target >= nums[start] && target < nums[mid] {
                end = mid - 1
            } else {
                start = mid + 1
            }
        } else {
            if target <= nums[end] && target > nums[mid] {
                start = mid + 1
            } else {
                end = mid - 1
            }
        }
    }
    return -1
}

File: 33. Search in Rotated Sorted Array/33.js
// 33. Search in Rotated Sorted Array solution in JavaScript
// https://leetcode.com/problems/search-in-rotated-sorted-array/description/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let low = 0;
    let high = nums.length - 1;

    while (low <= high) {
        let mid = low + Math.floor((high - low) / 2);

        if (nums[mid] == target) return mid;

        if (nums[low] <= nums[mid]) {
            if (nums[low] <= target && target <= nums[mid]) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        } else {
            if (nums[mid] <= target && target <= nums[high]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
    }

    return -1;
};